plugin/storage/grpc/shared/grpc_client.go

```diff
@@ -0,0 +1,179 @@
+	"github.com/jaegertracing/jaeger/storage/spanstore"
+)
+
+// GRPCClient implements shared.StoragePlugin and reads/writes spans and dependencies
+type GRPCClient struct {
+	readerClient     storage_v1.SpanReaderPluginClient
+	writerClient     storage_v1.SpanWriterPluginClient
+	depsReaderClient storage_v1.DependenciesReaderPluginClient
+}
+
+// GetTrace takes a traceID and returns a Trace associated with that traceID
+func (c *GRPCClient) GetTrace(ctx context.Context, traceID model.TraceID) (*model.Trace, error) {
+	stream, err := c.readerClient.GetTrace(ctx, &storage_v1.GetTraceRequest{
+		TraceID: traceID,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "plugin error")
+	}
+
+	trace := model.Trace{}
+	for {
+		received, err := stream.Recv()
+		if err == io.EOF {
+			break
+		}
+
+		if err != nil {
+			return nil, fmt.Errorf("stream error: %s", err)
+		}
+
+		for _, span := range received.Spans {
+			trace.Spans = append(trace.Spans, &span)
+		}
+	}
+
+	return &trace, nil
+}
+
+// GetServices returns a list of all known services
+func (c *GRPCClient) GetServices(ctx context.Context) ([]string, error) {
+	resp, err := c.readerClient.GetServices(ctx, &storage_v1.GetServicesRequest{})
+	if err != nil {
+		return nil, errors.Wrap(err, "plugin error")
+	}
+
+	return resp.Services, nil
+}
+
+// GetOperations returns the operations of a given service
+func (c *GRPCClient) GetOperations(ctx context.Context, service string) ([]string, error) {
+	resp, err := c.readerClient.GetOperations(ctx, &storage_v1.GetOperationsRequest{
+		Service: service,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "plugin error")
+	}
+
+	return resp.Operations, nil
+}
+
+// FindTraces retrieves traces that match the traceQuery
+func (c *GRPCClient) FindTraces(ctx context.Context, query *spanstore.TraceQueryParameters) ([]*model.Trace, error) {
+	stream, err := c.readerClient.FindTraces(context.Background(), &storage_v1.FindTracesRequest{
+		Query: &storage_v1.TraceQueryParameters{
+			ServiceName:   query.ServiceName,
+			OperationName: query.OperationName,
+			Tags:          query.Tags,
+			StartTimeMin:  query.StartTimeMin,
+			StartTimeMax:  query.StartTimeMax,
+			DurationMin:   query.DurationMin,
+			DurationMax:   query.DurationMax,
+			NumTraces:     int32(query.NumTraces),
+		},
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "plugin error")
+	}
+
+	var traces []*model.Trace
+	var trace *model.Trace
+	var traceID model.TraceID
+	for {
+		received, err := stream.Recv()
+		if err == io.EOF {
+			break
+		}
+
+		if err != nil {
+			return nil, errors.Wrap(err, "stream error")
+		}
+
+		for i, span := range received.Spans {
+			if span.TraceID != traceID {
+				if trace != nil {
+					traces = append(traces, trace)
+				}
+				trace = &model.Trace{}
+				traceID = span.TraceID
+			}
+			trace.Spans = append(trace.Spans, &received.Spans[i])
+		}
+	}
+	if trace != nil {
+		traces = append(traces, trace)
+	}
```

Comment:
rather than this, I'd change L120 to check for nil:
```
if trace == nil {
    trace = &model.Trace{}
    traceID = span.TraceID
    traces = append(traces, trace)
}
```

Reply:
The intention here is to catch the final trace as L119 onward won't get hit for that one. I don't think that this change would solve that problem?

Reply:
In my version we're adding new trace to the list as we create it and as we need it, so it doesn't make distinction between last / not last traces.